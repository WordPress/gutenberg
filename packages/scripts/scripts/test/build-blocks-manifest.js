/**
 * External dependencies
 */
const fs = require( 'fs' );
const path = require( 'path' );
const { execSync } = require( 'child_process' );
const rimraf = require( 'rimraf' );

const fixturesPath = path.join(
	__dirname,
	'fixtures',
	'build-blocks-manifest'
);
const outputPath = path.join( __dirname, 'build', 'test-blocks-manifest' );

describe( 'build-blocks-manifest script', () => {
	beforeAll( () => {
		if ( ! fs.existsSync( outputPath ) ) {
			fs.mkdirSync( outputPath, { recursive: true } );
		}
		rimraf.sync( outputPath );
	} );

	afterAll( () => {
		rimraf.sync( outputPath );
	} );

	it( 'should generate expected blocks manifest', () => {
		const inputDir = path.join( fixturesPath, 'input' );
		const outputFile = path.join( outputPath, 'blocks-manifest.php' );

		// Run the build-blocks-manifest script
		const scriptPath = path.resolve(
			__dirname,
			'..',
			'build-blocks-manifest.js'
		);
		execSync(
			`node ${ scriptPath } --input=${ inputDir } --output=${ outputFile }`
		);

		const generatedContent = fs.readFileSync( outputFile, 'utf8' );
		expect( generatedContent ).toMatchSnapshot();
	} );

	it( 'should error on empty input directory', () => {
		const emptyInputDir = path.join( fixturesPath, 'empty-input' );
		const outputFile = path.join( outputPath, 'empty-blocks-manifest.php' );

		const scriptPath = path.resolve(
			__dirname,
			'..',
			'build-blocks-manifest.js'
		);
		let error;
		try {
			execSync(
				`node ${ scriptPath } --input=${ emptyInputDir } --output=${ outputFile }`,
				{ encoding: 'utf8' }
			);
		} catch ( e ) {
			error = e;
		}

		// Check that an error was thrown.
		expect( error ).toBeDefined();
		expect( error.stdout ).toContain(
			`No block.json files were found in path`
		);

		// Ensure that the output file was not created
		expect( fs.existsSync( outputFile ) ).toBe( false );
	} );

	it( 'should error on missing input directory', () => {
		const nonExistentInputDir = path.join( fixturesPath, 'missing-input' );
		const outputFile = path.join( outputPath, 'empty-blocks-manifest.php' );

		const scriptPath = path.resolve(
			__dirname,
			'..',
			'build-blocks-manifest.js'
		);
		let error;
		try {
			execSync(
				`node ${ scriptPath } --input=${ nonExistentInputDir } --output=${ outputFile }`,
				{ encoding: 'utf8' }
			);
		} catch ( e ) {
			error = e;
		}

		// Check that an error was thrown.
		expect( error ).toBeDefined();
		expect( error.stdout ).toContain( `does not exist` );

		// Ensure that the output file was not created
		expect( fs.existsSync( outputFile ) ).toBe( false );
	} );
} );
