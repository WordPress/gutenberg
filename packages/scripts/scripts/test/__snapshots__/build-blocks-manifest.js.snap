// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build-blocks-manifest script should generate expected blocks manifest 1`] = `
"<?php
// This file is generated. Do not modify it manually.
return array(
	'custom-header' => array(
		'$schema' => 'https://schemas.wp.org/trunk/block.json',
		'apiVersion' => 2,
		'name' => 'my-plugin/custom-header',
		'title' => 'Custom Header',
		'category' => 'text',
		'icon' => 'heading',
		'description' => 'A custom header block with color options.',
		'attributes' => array(
			'content' => array(
				'type' => 'string',
				'source' => 'html',
				'selector' => 'h2'
			),
			'textColor' => array(
				'type' => 'string'
			),
			'backgroundColor' => array(
				'type' => 'string'
			)
		),
		'supports' => array(
			'html' => false,
			'color' => array(
				'background' => true,
				'text' => true
			)
		),
		'textdomain' => 'my-plugin',
		'editorScript' => 'file:./index.js',
		'editorStyle' => 'file:./index.css',
		'style' => 'file:./style-index.css'
	),
	'image-gallery' => array(
		'$schema' => 'https://schemas.wp.org/trunk/block.json',
		'apiVersion' => 2,
		'name' => 'my-plugin/image-gallery',
		'title' => 'Image Gallery',
		'category' => 'media',
		'icon' => 'format-gallery',
		'description' => 'An image gallery block with customizable layout.',
		'attributes' => array(
			'images' => array(
				'type' => 'array',
				'default' => array(
					
				),
				'source' => 'query',
				'selector' => 'img',
				'query' => array(
					'url' => array(
						'type' => 'string',
						'source' => 'attribute',
						'attribute' => 'src'
					),
					'alt' => array(
						'type' => 'string',
						'source' => 'attribute',
						'attribute' => 'alt',
						'default' => ''
					),
					'id' => array(
						'type' => 'number',
						'source' => 'attribute',
						'attribute' => 'data-id'
					)
				)
			),
			'columns' => array(
				'type' => 'number',
				'default' => 3
			),
			'imageCrop' => array(
				'type' => 'boolean',
				'default' => true
			)
		),
		'supports' => array(
			'align' => array(
				'wide',
				'full'
			)
		),
		'textdomain' => 'my-plugin',
		'editorScript' => 'file:./index.js',
		'editorStyle' => 'file:./index.css',
		'style' => 'file:./style-index.css'
	),
	'simple-button' => array(
		'$schema' => 'https://schemas.wp.org/trunk/block.json',
		'apiVersion' => 2,
		'name' => 'my-plugin/simple-button',
		'title' => 'Simple Button',
		'category' => 'design',
		'icon' => 'button',
		'description' => 'A simple button block.',
		'supports' => array(
			'html' => false
		),
		'textdomain' => 'my-plugin',
		'editorScript' => 'file:./index.js',
		'editorStyle' => 'file:./index.css',
		'style' => 'file:./style-index.css'
	)
);
"
`;
